lib/internal/abort_controller.js

```diff
@@ -0,0 +1,94 @@
+'use strict';
+
+// Modeled very closely on the AbortController implementation
+// in https://github.com/mysticatea/abort-controller (MIT license)
+
+const {
+  Object,
+  Symbol,
+} = primordials;
+
+const {
+  NodeEventTarget,
+  Event
+} = require('internal/event_target');
+const {
+  customInspectSymbol,
+  emitExperimentalWarning
+} = require('internal/util');
+const { inspect } = require('util');
+
+const kAborted = Symbol('kAborted');
+
+// Extends from NodeEventTarget intentionally to allow
+// the AbortSignal to be used similar to an EventEmitter.
+// This allows usage patterns like like:
+//   await once(abort.signal, 'abort')
+class AbortSignal extends NodeEventTarget {
+  get aborted() { return !!this[kAborted]; }
+
+  [customInspectSymbol](depth, options) {
+    const name = this.constructor.name;
+    if (depth < 0)
+      return name;
+
+    const opts = Object.assign({}, options, {
+      dept: options.depth === null ? null : options.depth - 1
+    });
+
+    return `${name} ${inspect({
+      aborted: this.aborted
+    }, opts)}`;
+  }
+}
+
+Object.defineProperties(AbortSignal.prototype, {
+  aborted: { enumerable: true }
+});
+
+class AbortEvent extends Event {
+  constructor() { super('abort'); }
+}
+
+function abortSignal(signal) {
+  if (signal[kAborted]) return;
+  signal[kAborted] = true;
+  const event = new AbortEvent();
+  if (typeof signal.onabort === 'function') {
+    signal.onabort(event);
+  }
+  signal.dispatchEvent(event);
+}
```

Comment:
Could we implement `abortSignal` so that it also removes all `abort` listeners afterwards? We know an AbortSignal can only be aborted once, so this would remove the need for `{ once: true }` and avoid a memory leak for anyone who forgot to use `{ once: true }` or the `onabort` setter.

Reply:
It doesn't seem like that would be spec-compliant, there's no mention of removing handlers after firing the event: https://dom.spec.whatwg.org/#abortsignal-signal-abort

Reply:
How would it be observable though? Since it's impossible to abort an already aborted `AbortController`?

I think if this is not done, this would be a major foot gun especially in the context of long-running Node servers, and it would really worry me to make it the blessed cancellation abstraction accepted throughout the Node stdlib with that foot gun. I've seen many people adopting `AbortSignal` and not using `once: true`.

Reply:
We cannot force `once` to be used if we are to remain spec compliant. We could potentially emit a warning but that's not ideal. In the docs for this, we should just show and strongly recommend using `once` and explain why

Reply:
Well... thinking about it further... there really is nothing in the AbortController spec that I can see that would forbid us from removing all event listeners after it triggers...

Reply:
I think removing listeners after firing the event is a good idea and to my understanding does not conflict with the spec as there is no spec aware way to observe it.

Reply:
@felixfbecker hmm, what if someone dispatches a second abort event through dispatchEvent on the signal?

Reply:
That's possible. But I wonder would you want to have your abort listener fired in that case? E.g. close that file handle that is already closed, abort that request that was already aborted, send another `shutdown` RPC message after one was already sent? All cases I can think of this would be undesirable and cause errors

Reply:
I'm more and more getting the feeling that `AbortSignal` being based on `EventTarget` is generally a bad abstraction... another point is that when you add a event handler after a Signal is already aborted, it will never get called (but it would be better if it was called immediately, like a Promise or Rx `Subscription`).

Maybe the way to go is to have Node export a static helper function `onceAbort()` and recommend to always interact with `AbortSignal`s through that? Which would set `{ once: true }` and call the handler immediately if the signal was already aborted.

Reply:
I think the value of AbortSignal is mostly in interoperability and as a tool for making universal JS easier. We should support a spec compliant AbortController if at all possible, but nothing is preventing us from shipping an extended version like NodeEventTarget/EventTarget

Reply:
The current behavior matches the browser implementations in Firefox and Chrome so I'm inclined to leave it as is. I did add a warning to the documentation to use `{ once: true }` to avoid the potential memory leak.